smartsim/experiment.py

```diff
@@ -276,6 +276,82 @@ def get_status(
         stats = (stats_map.get(i, InvalidJobStatus.NEVER_STARTED) for i in ids)
         return tuple(stats)
 
+    def wait(
+        self, *ids: LaunchedJobID, timeout: float | None = None, verbose: bool = True
+    ) -> None:
+        """Block execution until all of the provided launched jobs, represented
+        by an ID, have entered a terminal status.
+
+        :param ids: The ids of the launched ids to wait for.
+        :param timeout: The max time to wait for all of the launched jobs to end.
+        :param verbose: Whether found statuses should be displayed in the console.
+        :raises ValueError: No IDs were provided.
+        """
+        if not ids:
+            raise ValueError("No job ids to wait on provided")
+        self._poll_for_statuses(
+            ids,
+            TERMINAL_STATUSES,
+            timeout=_interval.SynchronousTimeInterval(timeout),
+            verbose=verbose,
+        )
+
+    def _poll_for_statuses(
+        self,
+        ids: t.Sequence[LaunchedJobID],
+        statuses: t.Collection[JobStatus],
+        timeout: _interval.SynchronousTimeInterval | None = None,
+        interval: _interval.SynchronousTimeInterval | None = None,
+        verbose: bool = True,
+    ) -> dict[LaunchedJobID, JobStatus | InvalidJobStatus]:
+        """Poll launchers until  status until all jobs represented by a
+        collections of ids have changed state to one of the provided statuses.
+
+        :param ids: IDs of launches to poll for status.
+        :param statuses: A collection of statuses to poll for.
+        :param timeout: The minimum amount of time to spend polling all jobs to
+            reach one of the supplied statuses. If not supplied or `None`, the
+            experiment will poll indefinitely.
+        :param interval: The minimum time between polling launchers.
+        :param verbose: Whether or not to log polled states the console.
+        :raises ValueError: The interval between polling launchers is infinite
+        :raises TimeoutError: The polling interval was exceeded.
+        :returns: A mapping of ids to the status they entered that ended
+            polling.
+        """
+        terminal = frozenset(itertools.chain(statuses, InvalidJobStatus))
+        log = logger.info if verbose else lambda *_, **__: None
+        method_timeout = timeout or _interval.SynchronousTimeInterval(None)
+        iter_timeout = interval or _interval.SynchronousTimeInterval(5.0)
```

Comment:
Should these be starting new intervals?
```suggestion
        method_timeout = (timeout or _interval.SynchronousTimeInterval(None)).new_interval()
        iter_timeout = (interval or _interval.SynchronousTimeInterval(5.0)).new_interval()
```
or is there value it letting someone pass in a partially elapsed time interval a-la
```py
t = _interval.SynchronousTimeInterval(10)
time.sleep(6)
exp._poll_for_statuses(ids, terminal_statuses, timeout=t)  # <- effectively now times out after 4 seconds
```
?

Reply:
That seems like it would get a little confusing for me (the partially elapsed time interval) Unless I am overthinking polling

Reply:
Parameter has been removed in favor of a simple `float` to avoid partial timeouts!